### PR TITLE
Allow dataloader iterator to be reused

### PR DESCRIPTION
**Pull request by zhreshold**
## Description ##
In gluon, DataLoader with very large ``num_workers`` can be slow at iteration initialization step, because forking a large number of processes can be expensive. 
This PR allow user to cache and reset to reuse the iterator by:

Updated:

```python

data_loader = mx.gluon.DataLoader(dataset, batch_size, num_workers=100)
it = iter(data_loader) # explicitly cache the iterator

for i in range(num_epochs):
    for batch in it:
        # training code
```

This PR don't affect the existing usage where each iterator is initialized and destroyed on-the-fly:
```python
data_loader = mx.gluon.DataLoader(dataset, batch_size, num_workers=100)
for i in range(epochs):
    for batch in data_loader:
        # training code
```
  

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by zhreshold** [comment]
@hetong007 @szha for review.

File: python/mxnet/gluon/data/dataloader.py [hunk 1]
```diff
@@ -244,6 +270,35 @@ def __len__(self):
     def __del__(self):
         self.shutdown()
 
+    def reset(self):
```
Comment (szha) [inline python/mxnet/gluon/data/dataloader.py]: assert not shutdown

File: python/mxnet/gluon/data/dataloader.py [hunk 2]
```diff
@@ -281,6 +335,58 @@ def shutdown(self):
...
+        If ``True``, the dataloader will copy NDArrays into pinned memory
+        before returning them. Copying from CPU pinned memory to GPU is faster
+        than from normal CPU memory.
+
+    """
+    def __init__(self, dataset, batchify_fn, batch_sampler, pin_memory=False):
+        self._dataset = dataset
+        self._batchify_fn = batchify_fn
+        self._batch_sampler = batch_sampler
+        self._iter = iter(self._batch_sampler)
```
Comment (szha) [inline python/mxnet/gluon/data/dataloader.py]: call reset here in case more states need to be maintained in reset()

File: python/mxnet/gluon/data/dataloader.py (hunk 1 above)
Comment (leezu) [reply python/mxnet/gluon/data/dataloader.py]: Do we need to introduce a new reset method? Just moving this functionality into `__iter__` of this class would be cleaner I think

**Comment by anirudhacharya** [comment]
@mxnet-label-bot add [pr-awaiting-review]

**Comment by leezu** [comment]
Thanks for your work @zhreshold . I think you can just remove the `reset()` method. The logic should instead be handled in `_MultiWorkerIter.__iter__` and `_SameProcessIter.__iter__`. That would simplify the use case of your example:

```
data_loader = mx.gluon.DataLoader(dataset, batch_size, num_workers=100)
it = iter(data_loader) # explicitly cache the iterator

for i in range(num_epochs):
    # it.reset() # Deleted. Not necessary anymore.
    for batch in it: # This calls it.__iter__ which resets the iterator
        # training code
```

Still does not affect the existing use-case:

```
data_loader = mx.gluon.DataLoader(dataset, batch_size, num_workers=100)
for i in range(epochs):
    for batch in data_loader:
        # training code
```

However, I wonder if there is a cleaner way to represent the process pool. The current solution may be a bit confusing to users. It is not very intuitive that the iterator must be cached for good performance. Explicitly specifying the resource (ie. process pool) used for the execution may help to clarify? Are you familiar with the `concurrent.futures` pools? I feel like we are reimplementing a slightly different way of this Python standard libray API here. It may be better to find abstractions that just rely on the Python standard API. It may also be easier to understand for new users? For example, the concept of a executor may be useful. For multiprocessing, there is https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor

The last paragraph is not a critic of your PR but I want to raise the point that a overhaul of the API may help improve clarity. We can introduce a dependency on the [backport of `concurrent.futures`](https://www.google.com/search?q=concurrent.futures+python+2&ie=utf-8&oe=utf-8&client=firefox-b-ab) to Py2 for Python 2 users (Python 2 is EOL soon). @szha

**Comment by zhreshold** [comment]
@leezu Good point, using __iter__ is definitely much cleaner in such use cases. Let me go back to the original concerns of pool, and will update later.

**Comment by leezu** [comment]
@zhreshold ok, great. Btw, I noticed you missed updating `_SameProcessIter` regarding `reset / __iter__` in your last commit. Thanks for your work on this, I think addressing speed concerns in the Gluon API is very important for a good user experience.

**Comment by stu1130** [comment]
@zhreshold could you address the CI failure
@szha ping for the second round review

**Comment by leezu** [comment]
@stu1130 I think @zhreshold still plans to make some changes (my understanding of https://github.com/apache/incubator-mxnet/pull/13209#issuecomment-437983209) If so, no need for review yet

**Comment by zhreshold** [comment]
@leezu I currently do have an option to utilize process pools, and it should work elegantly without caching iterators explicitly. 
However, due to the backward compatibility issue of future module mentioned here https://pypi.org/project/futures/:
*"The ProcessPoolExecutor class has known (unfixable) problems on Python 2 and should not be relied on for mission critical work."* 

Therefore I am not including the update in this PR. Besides, the logic to protect RecordIO handles is moved to `recordio.py`. This PR can improve initialization time significantly (more than 5x) on average according to local benchmark.

Please have a second look @leezu @szha @hetong007

File: python/mxnet/gluon/data/dataloader.py [hunk 3]
```diff
@@ -363,13 +449,8 @@ def __init__(self, dataset, batch_size=None, shuffle=False, sampler=None,
 
     def __iter__(self):
```
Comment (leezu) [inline python/mxnet/gluon/data/dataloader.py]: Would it make sense to add documentation about the API of returned `_SameProcessIter` / `_MultiWorkerIter`? Ie. that they will reset when `__iter__` is called on them.

File: python/mxnet/gluon/data/dataloader.py [hunk 4]
```diff
@@ -203,7 +187,29 @@ def fetcher_loop(data_queue, data_buffer, pin_memory=False):
...
+        Callback function to allow users to specify how to merge samples
+        into a batch.
+    batch_sampler : Sampler
+        A sampler that returns mini-batches. Do not specify batch_size,
+        shuffle, sampler, and last_batch if batch_sampler is specified.
+    pin_memory : boolean, default False
+        If ``True``, the dataloader will copy NDArrays into pinned memory
+        before returning them. Copying from CPU pinned memory to GPU is faster
+        than from normal CPU memory.
+
```
Comment (eric-haibin-lin) [inline python/mxnet/gluon/data/dataloader.py]: Can we also document `worker_fn`?

File: python/mxnet/gluon/data/dataloader.py [hunk 5]
```diff
@@ -267,20 +316,128 @@ def __next__(self):
                 return batch
 
     def next(self):
+        """Compatible portal for __next__ in python2.
+
+        Returns
+        -------
+        type
+            Description of returned object.
```
Comment (eric-haibin-lin) [inline python/mxnet/gluon/data/dataloader.py]: need to update doc

File: python/mxnet/gluon/data/dataloader.py [hunk 6]
```diff
@@ -267,20 +316,128 @@ def __next__(self):
...
+        alive, thus save re-init time of forking processes.
+
+        Returns
+        -------
+        iterator
+            Iterator of self.
+
+        """
+        self._reset()
         return self
```
Comment (eric-haibin-lin) [inline python/mxnet/gluon/data/dataloader.py]: What happens if users used the dataloader and created two iterators? Does the 2nd iterator reset the state for the first one that is running? For example:
```
>>> a = [1,2,3]
>>> for x in a:
...     for y in a:
...         print(x,y)
...
(1, 1)
(1, 2)
(1, 3)
(2, 1)
(2, 2)
(2, 3)
(3, 1)
(3, 2)
(3, 3)
```

Comment (zhreshold) [reply python/mxnet/gluon/data/dataloader.py]: iterators from the same dataloader are two separate instances, they don't interfere with each other, however, for explicitly cached iterator, it will break

```
a = [1,2,3]
it = iter(a)
for x in it:
    for y in it:
        # this will not work

File: python/mxnet/gluon/data/dataloader.py (hunk 5 above)
Comment (zhreshold) [reply python/mxnet/gluon/data/dataloader.py]: will do

File: python/mxnet/gluon/data/dataloader.py [hunk 7]
```diff
@@ -203,7 +187,29 @@ def fetcher_loop(data_queue, data_buffer, pin_memory=False):
 
 
 class _MultiWorkerIter(object):
-    """Interal multi-worker iterator for DataLoader."""
+    """Interal multi-worker iterator for DataLoader.
```
Comment (anirudhacharya) [inline python/mxnet/gluon/data/dataloader.py]: Interal -> Internal

**Comment by zhreshold** [comment]
Close this in favor of better solution: #13447
